### PR TITLE
sql_instance: migrate to rbr compatible index

### PR DESCRIPTION
**Pull request by jeffswenson**
Migrate sql_instance to a regional by row compatible index. The version
gates are intended to follow the protocol discussed in the comment at
the top of upgrades/system_rbr_indexes.go

The crdb_region column ID was changed from 5 to 6 in order to match the
logical order in which the sql_addr and crdb_region columns were added.
The exact ID doesn't really matter in this case since the sql_addr
column was added in v23.1.

Most of the rbr migration work is the same for sqlliveness, lease, and
sql_instances. The main exception to that is the migration cache used by
the sql instance reader. The cache is backed by a range feed and we need
to switch implementations when the version setting changes.

Part of #94843

Relase note: None

### CODE REVIEW

**Review by ajwerner — COMMENTED** [review]
This is close. Some of the logic is quite subtle and deserves a bit more commentary and another pass to try to make the control flow clearer. 

> Reviewed 13 of 22 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98445#-:-NQRvwmL-DzNVyqUfELW:b-jwoaco)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @JeffSwenson)

___
*[`pkg/sql/sqlinstance/instancestorage/instancecache.go` line 223 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/98445#-NQRslDN96SsqLGRh-Tp:-NQRslDN96SsqLGRh-Tq:b-ndqi1l) ([raw file](https://github.com/cockroachdb/cockroach/blob/df42bbacfb9ee782a4a33827ab40af2ce659f0b5/pkg/sql/sqlinstance/instancestorage/instancecache.go#L223)):*
> ```Go
> // from the old implementation to the new implementation when the version
> // changes to V23_1_SystemRbrReadNew.
> func newMigrationCache(
> ```

this function is a beast. 
___
*[`pkg/sql/sqlinstance/instancestorage/instancecache.go` line 253 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/98445#-NQRrJzb66ZIujRiMBMH:-NQRrJzb66ZIujRiMBMI:b9v8342) ([raw file](https://github.com/cockroachdb/cockroach/blob/df42bbacfb9ee782a4a33827ab40af2ce659f0b5/pkg/sql/sqlinstance/instancestorage/instancecache.go#L253)):*
> ```Go
> 		// Use the background context because we don't want RPC cancellation to
> 		// prevent starting the new range feed.
> 		err := stopper.RunAsyncTask(context.Background(), "start-new-cache-implementation", func(ctx context.Context) {
> ```

nit: annotate the context with logtags. Probably `cacheCtx := logtags.AddTags(context.Background(), logtags.FromContext(ctx))`
___
*[`pkg/sql/sqlinstance/instancestorage/instancecache.go` line 256 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/98445#-NQRuHUa6LP3lx5VsQ_6:-NQRuHUa6LP3lx5VsQ_7:bjiwkyd) ([raw file](https://github.com/cockroachdb/cockroach/blob/df42bbacfb9ee782a4a33827ab40af2ce659f0b5/pkg/sql/sqlinstance/instancestorage/instancecache.go#L256)):*
> ```Go
> 			log.Ops.Info(ctx, "starting new system.sql_instance cache")
> 
> 			cache, err := newCache(ctx)
> ```

I'd feel better if there were a defer which sent on this channel the return value of a function. In general, I feel like it'd be nicer if you made closures for two functions which send on the channels and lay out the invariants about who is going to send on which channel. After having scrutinized this a bunch, I don't see obvious bugs if there are no panics, but this thing is surely complex.
___
*[`pkg/sql/sqlinstance/instancestorage/instancecache.go` line 306 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/98445#-NQRuiTE59hxvX_RBDU-:-NQRuiTE59hxvX_RBDU0:bikmrm9) ([raw file](https://github.com/cockroachdb/cockroach/blob/df42bbacfb9ee782a4a33827ab40af2ce659f0b5/pkg/sql/sqlinstance/instancestorage/instancecache.go#L306)):*
> ```Go
> 
> 	select {
> 	case err = <-newReady:
> ```

I think there's a subtle race here. Imagine that the version change notification happens and then the stopper gets shut down. There's a case where the `versionChanged` field could get set to true, and the `RunAsyncTask` code can choose not to run. We'll log a warning but nobody will send on either channel. I think you can fix that by sending on the channel when unable to start the new cache. 

All of this code is pretty subtle. My sense is that splitting out some of the logic into helpers can make the behavior a bit clearer. 
___
*[`pkg/sql/tests/system_table_test.go` line 223 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/98445#-NQRvuGuAxDrb6xxd1c5:-NQRvuGuAxDrb6xxd1c6:b54vkmo) ([raw file](https://github.com/cockroachdb/cockroach/blob/df42bbacfb9ee782a4a33827ab40af2ce659f0b5/pkg/sql/tests/system_table_test.go#L223)):*
> ```Go
> 		require.NoError(t, desctestutils.TestingValidateSelf(gen))
> 
> 		switch gen.GetID() {
> ```

Explain this in commentary

**Review by jeffswenson — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98445)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @yuzefovich)

___
*[`pkg/sql/sqlinstance/instancestorage/instancecache.go` line 223 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/98445#-NQRslDN96SsqLGRh-Tp:-NQVOjaBA1Jw6i7mJwJC:birjlbk) ([raw file](https://github.com/cockroachdb/cockroach/blob/df42bbacfb9ee782a4a33827ab40af2ce659f0b5/pkg/sql/sqlinstance/instancestorage/instancecache.go#L223)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

this function is a beast. 
</blockquote></details>

I reworked the function and tried to clean up some of the state. Let me know what you think.
___
*[`pkg/sql/sqlinstance/instancestorage/instancecache.go` line 256 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/98445#-NQRuHUa6LP3lx5VsQ_6:-NQVP7Y5DVrbVYFmQZyK:b3dcla0) ([raw file](https://github.com/cockroachdb/cockroach/blob/df42bbacfb9ee782a4a33827ab40af2ce659f0b5/pkg/sql/sqlinstance/instancestorage/instancecache.go#L256)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I'd feel better if there were a defer which sent on this channel the return value of a function. In general, I feel like it'd be nicer if you made closures for two functions which send on the channels and lay out the invariants about who is going to send on which channel. After having scrutinized this a bunch, I don't see obvious bugs if there are no panics, but this thing is surely complex.
</blockquote></details>

I've reworked the implementation. Let me know if it is still difficult to follow and I will take another crack at it.
___
*[`pkg/sql/sqlinstance/instancestorage/instancecache.go` line 306 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/98445#-NQRuiTE59hxvX_RBDU-:-NQVPLSo0_0JHFXvEm4s:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/df42bbacfb9ee782a4a33827ab40af2ce659f0b5/pkg/sql/sqlinstance/instancestorage/instancecache.go#L306)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I think there's a subtle race here. Imagine that the version change notification happens and then the stopper gets shut down. There's a case where the `versionChanged` field could get set to true, and the `RunAsyncTask` code can choose not to run. We'll log a warning but nobody will send on either channel. I think you can fix that by sending on the channel when unable to start the new cache. 

All of this code is pretty subtle. My sense is that splitting out some of the logic into helpers can make the behavior a bit clearer. 
</blockquote></details>

Done.
___
*[`pkg/sql/tests/system_table_test.go` line 223 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/98445#-NQRvuGuAxDrb6xxd1c5:-NQVPiTYAvWh0xRFqnYA:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/df42bbacfb9ee782a4a33827ab40af2ce659f0b5/pkg/sql/tests/system_table_test.go#L223)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Explain this in commentary
</blockquote></details>

Done.

File: pkg/sql/sqlinstance/instancestorage/row_codec_test.go [hunk 1]
```diff
@@ -40,11 +40,11 @@ func TestRowCodec(t *testing.T) {
 
 	t.Run("RegionalByRow", func(t *testing.T) {
```
Comment (ajwerner) [inline pkg/sql/sqlinstance/instancestorage/row_codec_test.go:41]: nit: see `testutils.RunTrueAndFalse`?

**Review by ajwerner — APPROVED** [review]
LGTM

File: pkg/sql/sqlinstance/instancestorage/row_codec_test.go (hunk 1 above)
Comment (JeffSwenson) [reply pkg/sql/sqlinstance/instancestorage/row_codec_test.go]: Done

**Pull request merged**